README.md
[![Crate release version](https://flat.badgen.net/crates/v/passcod-noodle)](https://crates.io/crates/passcod-noodle)
[![Crate license: Apache 2.0 or MIT](https://flat.badgen.net/badge/license/Apache%202.0%20or%20MIT)][copyright]
![MSRV: latest stable](https://flat.badgen.net/badge/MSRV/latest%20stable/orange)
[![Uses Caretaker Maintainership](https://flat.badgen.net/badge/Caretaker/Maintainership%20👥%20/purple)][caretaker]

# Noodle

_AKA "someone stop Félix from naming things, please, this is terrible."_

- Go directly to the [usage summary](#obtain).
- [Dual-licensed][copyright] with Apache 2.0 and MIT.
- Uses [Caretaker Maintainership][caretaker].

[caretaker]: ./CARETAKERS.md
[copyright]: ./COPYRIGHT

## Introduction

Noodle as in pool noodle, because it floats and IPs and... geddit?

The idea is based on [MetalLB](https://metallb.universe.tf/concepts/layer2/).

Basically, we emit [RFC5944 "gratuitous" ARP announcements](https://tools.ietf.org/html/rfc5944#section-4.6).

Instead of the usual ARP request-reply "who has 1.2.3.4?" "I do" cycle, we send
"I have 1.2.3.4" and we add 1.2.3.4 to the network interface we do that from,
so the network stack handles traffic sent to it properly.

Noodle is the little daemon that does the ARP announcements like that.

But Noodle also listens for ARP on the interface. If it sees ARP on the
interface for the same IP it's announcing, but with a different MAC address, it
stops and exits!

When it stops, if it's being managed by an active orchestrator or supervisor,
it will get restarted. And if it's not, but for some reason wasn't shut down
when the orchestrator or supervisor went down, it wont come back up.

In any case, an active Noodle will keep spraying its ARP over the subnet
every N seconds, so well behaved devices like routers and VMs will keep
their ARP tables updated with the MAC of whatever device Noodle is on.

Essentially it's like MetalLB but we use the network itself as an additional
consensus layer, and the main consensus is via whatever orchestrator you use.
You can also disable the ARP watcher and/or only send a number of ARPs and then
quit, so you can build the orchestration a little more granularly yourself if
you want. It's up to you!

BTW this was written for use with Nomad but doesn't in any way depend on Nomad.

## Good times

### In a graceful move situation

Orchestrator will boot Noodle on the other Node, and then either the
orchestrator or the new ARP will kill the old Noodle. ARP tables will be
updated subnet-wide and most importantly at the router, and the IP will
essentially have "floated" over to the new node in short order.

The orchestrator, when configured properly, can also keep the old service
running for a bit before killing it even while starting the new one, so traffic
will move over as smoothly as it can (long connections may still get broken).

### In a failover situation

The old node will be dead. Traffic goes nowhere. Devices on the subnet
and the router still have the old MAC in the ARP tables.

Orchestrator notices a node is down, and reschedules its workload on the alive
nodes. Noodle starts, blathers ARP announcements over the network, tables get
updated, and traffic starts flowing again.

### In a split brain situation

Orchestrator worker node gets separated from the orchestrator's servers, or
talks to only one server and _it_ is separated from the others. In either case,
that node's orchestrator worker agent will:

- appear dead to the rest of the cluster, which will reschedule work as
  in the above, and
- notice that it's dead or doesn't have consensus, and _ideally_ will
  kill itself.

Proceeds mostly the same as failover, but may proceed as crash:

### In an orchestrator crash situation

Orchestrator crashes, leaving the underlying containers or applications
running. Rest of the cluster notices the node is dead and reschedules. The new
Noodle yells its announce, the old Noodle notices ARP that is not coming from
itself, and exits. Traffic moves over to the new node.

## Bad times

### Partial split brain where the old node still thinks it's meant to be up

Old orchestrator appears dead to the rest of the cluster, which reschedules.

Old Noodle notices ARP, and kills itself.

Old orchestrator restarts its Noodle.

New Noodle notices ARP, and kills itself.

New orchestrator restarts its Noodle.

Round and round it goes, traffic is mightily confused, routers might
take a hit.

### In a Noodle crash situation

Noodle crashes in the middle of its loops and doesn't clean up the IP on
its interface. Linux responds to ARP requests with solicited "it's me!"
while other Noodle is screaming out "it's me it's me it's me" announces.

Traffic and routers get confused.

_This is always a bug and should be reported._

### Operator accidentally scales the Noodle service for one floating IP to 2 or more instances

Proceeds like partial split brain except both nodes are technically legitimate.

## Alternatives

- Native Cloud Provider Floating IP: absolutely use that if available.

- Tunnels: requires router access to set up and tunnel endpoint at the
  service, can be connected to twice concurrently and then what. Adds a
  moving piece. But is much quieter on the network, cleaner
  traffic-wise, maybe clearer to netops.

- BGP: requires router access to set up, and complicated daemon on every
  node, plus management of said daemons' config. Does offer true IP-level
  load-balancing instead of just directing traffic to one host. BGP can
  be a pain and also may interfere with netops.

- DNS: externalises the problem, very slow to update even with TTL=0 because of
  client behaviour. Some traffic just never dies.

- Single VM acts as load-balancer and is never rebooted: not really an
  option, this is what we're trying to get away from.

- Transferring the entire NIC at the VM hypervisor level between machines
  instead of doing the ARP dance: certainly an option, relies on having
  hypervisor API access, needs some kind of centralised control to avoid
  split brain and conflicts. Implies downtime during the switching.

- Just changing the hypervisor MAC of the interfaces: same as previous,
  plus adds even more downtime as MACs can't conflict across VMs and
  it needs to find a free "transfer" mac or shut the old NIC down.

- Changing the ARP table of the router directly with the router's
  administrative API: doesn't yell so much into the void, but requires access
  to router, and some kind of synchronisation to avoid conflicting commands.

- VRRP, Tree Spanning, OSPF, bonding, etc: cool tech, may work, requires
  netops. May not handle intra-subnet traffic, only via router.

## Obtain

Only works on Linux.

Currently only ARP (supporting IPv4) is implemented.

### From binary release

The [release tab on GitHub](https://github.com/passcod/noodle/releases).

Builds are available for:

- x86-64, both gnu and musl
- AArch64, both gnu and musl
- Arm7 HF, both gnu and musl

It's trivial to add more, so please ask.

### ~~With cargo binstall~~

⚠  Not available yet, depends on [netlink#149](https://github.com/little-dude/netlink/issues/149)

```
cargo binstall passcod-noodle
```

### ~~From source~~

⚠  Not available yet, depends on [netlink#149](https://github.com/little-dude/netlink/issues/149)

```
cargo install passcod-noodle
```

You can also compile from the repo as usual.

## Use

Requires sudo or the `NET_ADMIN` capability.

Minimal command:

```
noodle --ip 10.9.8.7/24 --interface ens123
```

Mandatory options:

- `--ip IP/SUBNET`: the floating IP to announce.
- `--interface NAME`: which interface to announce ARP on.

Other options:

- `--mac ADDRESS`: override the MAC address IP is announced for (default=read from interface).
- `--target ADDRESS`: override the MAC address packets are sent to (default=broadcast).
- `--log LEVEL`: specify the log level (default=info). All logs are JSON.
- `--interval DURATION` in seconds (default=10): how often to announce.
- `--delay DURATION` in seconds (default=0): delay the first announce.
- `--watch-delay DURATION` in seconds (default=0): delay before watching for competings.
- `--jitter DURATION` in seconds (default=1): add jitter to each interval up to this value.
- `--arp-reply`: use ARP reply instead of ARP request as announcement type.
- `--unmanaged-ip`: leave the interface alone (don't add/remove the ip).
- `--die-if-ip-exists`: exit with status 1 if the IP already exists on the interface.
- `--remove-pre-existing-ip`: remove the IP even if we didn't add it ourselves.
- `--watch BEHAVIOUR`: control the competing announcement watcher:
  * `fail` (default): exit with status 1 if we see an announcement for this
    IP by another MAC address
  * `quit`: exit with status 0 instead
  * `log`: don't exit, only log it
  * `no`: don't watch
- `--watch-immediately`: don't wait until the first announce is sent to start watching.
- `--count N` (default=0/disabled): only announce this many times.
- `--once`: shorthand for `--count 1  --delay 0  --jitter 0  --watch no`.

Info switches:

- `--help`: print the help.
- `--readme`: print this readme.
- `--source`: print the source (Cargo.toml and main.rs).
- `--version`: print the version number.


CHANGELOG.md
# Changelog

## v1.1.1 (2021-03-22)

- Fix bug where watcher would always start immediately if `--count` wasn't provided.

## v1.1.0 (2021-03-22)

- Don't apply jitter to delays.
- Add `--watch-delay` option.
- Add `--watch-immediately` option.
- Wait until the first announce is sent before starting the competing announce watcher.
- Don't crash if the IP exists on the interface already.
- Add `--die-if-ip-exists`.
- Add `--remove-pre-existing-ip`.
- Add warning if a /32 is used.

## v1.0.1 (2021-03-21)

- Handle runtime (as opposed to startup time) errors with logging rather than eyre.
- Remove readme mention of error code 17 (was never true).
- Actually exit with code 1 after a runtime error (was 0).
- Retire debug mode pretty logger.
- Omit other modules from logs except at trace level.
- Add module path to logs at trace level.
- Use string level instead of bunyan integer level.
- Add timestamps to logs.

## v1.0.0 (2021-03-20)

Initial release


src/main.rs
// SPDX-License-Identifier: Apache-2.0 OR MIT

use std::{
	convert::TryFrom,
	io::Write,
	net::{IpAddr, Ipv4Addr, Ipv6Addr},
	process::exit,
	result::Result as StdResult,
	str::FromStr,
	thread::sleep,
	time::Duration,
};

use argh::FromArgs;
use async_std::{
	channel,
	prelude::FutureExt,
	task::{block_on, spawn, spawn_blocking},
};
use chrono::{SecondsFormat, Utc};
use color_eyre::eyre::{eyre, Result};
use env_logger::{Builder as LogBuilder, Target as LogTarget};
use futures::{stream::TryStreamExt, TryFutureExt};
use kv_log_macro::{debug, error, info, warn};
use log::{kv, LevelFilter};
use pnet::{
	datalink::{
		channel as datachannel, interfaces, Channel, ChannelType, Config, NetworkInterface,
	},
	ipnetwork::IpNetwork,
	packet::{
		arp::{
			ArpHardwareType, ArpHardwareTypes, ArpOperation, ArpOperations, ArpPacket,
			MutableArpPacket,
		},
		ethernet::{EtherTypes, EthernetPacket, MutableEthernetPacket},
		MutablePacket, Packet,
	},
	util::MacAddr,
};
use pulse::Signal;
use rand::{rngs::OsRng, Rng};
use rtnetlink::{packet::{AddressMessage, rtnl::address::nlas::Nla}, AddressHandle};
use serde::Serialize;

macro_rules! as_display {
	($e:expr) => {
		&$e as &dyn std::fmt::Display
	};
}

const SOURCE_MAIN: &str = include_str!("main.rs");
const SOURCE_CARGO: &str = include_str!("../Cargo.toml");
const README: &str = include_str!("../README.md");

/// Announce an IP for an interface via ARP.
/// If we hear someone else announcing the IP, stop.
/// See more details in README. License: Apache 2.0 OR MIT.
#[derive(Debug, FromArgs)]
struct Args {
	/// interface to announce on (required)
	#[argh(option)]
	interface: Option<String>,

	/// ip in address/subnet format to announce (required)
	#[argh(option)]
	ip: Option<IpNetwork>,

	/// mac address override
	#[argh(option)]
	mac: Option<MacAddr>,

	/// target mac override (default=broadcast)
	#[argh(option, default = "MacAddr::broadcast()")]
	target: MacAddr,

	/// log level (default=info)
	///
	/// [no, error, warn, info, debug, trace]
	#[argh(option, default = "Default::default()")]
	log: LogLevel,

	/// interval in seconds between announcements (default=10)
	#[argh(option, from_str_fn(str_to_secs), default = "Duration::from_secs(10)")]
	interval: Duration,

	/// delay in seconds before sending first announce (default=0/disabled)
	#[argh(option, from_str_fn(str_to_secs), default = "Duration::from_secs(0)")]
	delay: Duration,

	/// delay in seconds before watching for competing announcements (default=0/disabled)
	#[argh(option, from_str_fn(str_to_secs), default = "Duration::from_secs(0)")]
	watch_delay: Duration,

	/// add some random [0 - value in seconds] jitter to each interval (default=1)
	#[argh(option, from_str_fn(str_to_secs), default = "Duration::from_secs(1)")]
	jitter: Duration,

	/// announce this many times then stop (default=0/disabled)
	#[argh(option, default = "0")]
	count: usize,

	/// control what the competing announcement watcher does when it sees ARP for the same IP but
	/// from a different MAC (default=fail)
	///
	/// [fail: exit with code=1]
	/// [quit: exit with code=0]
	/// [log: don't exit, only log]
	/// [no: don't watch]
	#[argh(option, default = "Default::default()")]
	watch: Watch,

	/// start the watcher immediately instead of waiting until the first announcement
	#[argh(switch)]
	watch_immediately: bool,

	/// use arp reply instead of request announcements
	#[argh(switch)]
	arp_reply: bool,

	/// don't add/remove the ip to/from the interface
	#[argh(switch)]
	unmanaged_ip: bool,

	/// exit with code=1 if the ip exists on the interface already
	#[argh(switch)]
	die_if_ip_exists: bool,

	/// remove the ip from the interface on exit even when we didn't add it ourselves
	#[argh(switch)]
	remove_pre_existing_ip: bool,

	/// shorthand for `--delay=0 --jitter=0 --count=1 --watch=no`
	#[argh(switch)]
	once: bool,

	/// print the source
	#[argh(switch)]
	source: bool,

	/// print the readme
	#[argh(switch)]
	readme: bool,

	/// print the version
	#[argh(switch)]
	version: bool,
}

fn str_to_secs(s: &str) -> StdResult<Duration, String> {
	u64::from_str(s)
		.map_err(|e| e.to_string())
		.map(Duration::from_secs)
}

#[derive(Clone, Copy, Debug)]
enum Watch {
	Fail,
	Quit,
	Log,
	No,
}

impl Default for Watch {
	fn default() -> Self {
		Self::Fail
	}
}

impl FromStr for Watch {
	type Err = String;

	fn from_str(s: &str) -> StdResult<Self, Self::Err> {
		match s.to_ascii_lowercase().as_str() {
			"fail" => Ok(Self::Fail),
			"quit" => Ok(Self::Quit),
			"log" => Ok(Self::Log),
			"no" => Ok(Self::No),
			_ => Err(String::from("invalid --watch value")),
		}
	}
}

#[derive(Clone, Copy, Debug)]
enum LogLevel {
	No,
	Error,
	Warn,
	Info,
	Debug,
	Trace,
}

impl Default for LogLevel {
	fn default() -> Self {
		Self::Info
	}
}

impl FromStr for LogLevel {
	type Err = String;

	fn from_str(s: &str) -> StdResult<Self, Self::Err> {
		match s.to_ascii_lowercase().as_str() {
			"no" | "none" => Ok(Self::No),
			"error" => Ok(Self::Error),
			"warn" | "warning" => Ok(Self::Warn),
			"info" => Ok(Self::Info),
			"debug" | "verbose" => Ok(Self::Debug),
			"trace" => Ok(Self::Trace),
			_ => Err(String::from("invalid --log value")),
		}
	}
}

impl LogLevel {
	fn install(self) -> Result<()> {
		if let Self::No = self {
			return Ok(());
		}

		let mut log = LogBuilder::new();
		log.target(LogTarget::Stdout);

		if let Self::Trace = self {
			log.filter(None, LevelFilter::Trace);
		} else {
			log.filter(
				Some("noodle"),
				match self {
					Self::Error => LevelFilter::Error,
					Self::Warn => LevelFilter::Warn,
					Self::Info => LevelFilter::Info,
					Self::Debug => LevelFilter::Debug,
					_ => unreachable!(),
				},
			);
		}

		#[derive(Serialize)]
		struct Record<'kv> {
			level: &'static str,
			#[serde(skip_serializing_if = "Option::is_none")]
			module: Option<String>,
			ts: String,
			msg: String,

			#[serde(flatten)]
			#[serde(with = "kv::source::as_map")]
			kvs: &'kv dyn kv::Source,
		}

		log.format(move |mut buf, record| {
			let rec = Record {
				level: record.level().as_str(),
				module: if let Self::Trace = self {
					record.module_path().map(|m| m.to_string())
				} else {
					None
				},
				ts: Utc::now().to_rfc3339_opts(SecondsFormat::Millis, true),
				msg: record.args().to_string(),
				kvs: record.key_values(),
			};

			serde_json::to_writer(&mut buf, &rec)?;
			writeln!(buf)?;
			Ok(())
		});

		log.try_init()?;

		Ok(())
	}
}

fn wait(d: Duration) {
	if d.as_millis() > 0 {
		debug!("sleeping {:?}", d);
		sleep(d);
	}
}

fn jittered(base: Duration, jitter: Duration) -> Duration {
	match (base.as_secs(), jitter.as_millis()) {
		(0, 0) => Duration::from_secs(0),
		(_, 0) => base,
		(_, j) => {
			base + Duration::from_millis(OsRng::default().gen_range(0..u64::try_from(j).unwrap()))
		}
	}
}

fn main() -> Result<()> {
	// panics+prep errors get color-eyre'd, run errors get logged
	color_eyre::install()?;

	if let Some(p) = prep()? {
		debug!("arguments", {
			ip: &p.0.to_string(),
			interface: &p.1.to_string(),
			mac: &p.2.to_string(),
			ip_managed: p.3,
			args: &format!("{:?}", p.4),
		});

		if let Err(e) = block_on(run(p)) {
			error!("{}", e);
			exit(1);
		}
	}

	Ok(())
}

type Prep = (IpNetwork, NetworkInterface, MacAddr, bool, Args);

fn prep() -> Result<Option<Prep>> {
	let (ip, args) = {
		let mut args: Args = argh::from_env();
		args.log.install()?;

		if args.source {
			println!(
				"# Cargo.toml\n{}\n\n# src/main.rs\n{}",
				SOURCE_CARGO, SOURCE_MAIN
			);
			return Ok(None);
		}

		if args.readme {
			println!("{}", README);
			return Ok(None);
		}

		if args.version {
			println!("{}", env!("CARGO_PKG_VERSION"));
			return Ok(None);
		}

		if args.once {
			args.delay = Duration::from_secs(0);
			args.jitter = Duration::from_secs(0);
			args.count = 1;
			args.watch = Watch::No;
		}

		if args.jitter > args.interval {
			return Err(eyre!("jitter > interval makes no sense"));
		}

		if args.delay > Duration::from_secs(60 * 60 * 24) {
			warn!("delay > 24h is probably a mistake");
		}

		if args.interval > Duration::from_secs(60 * 60 * 24) {
			warn!("interval > 24h is probably a mistake");
		}

		match (args.ip, &args.interface) {
			(Some(ip), Some(_)) => (ip, args),
			(Some(_), None) => return Err(eyre!("missing required option: --interface")),
			(None, Some(_)) => return Err(eyre!("missing required option: --ip")),
			(None, None) => return Err(eyre!("missing required options: --interface, --ip")),
		}
	};

	let one_prefix = match ip {
		IpNetwork::V4(_) => 32,
		IpNetwork::V6(_) => 128,
	};

	if ip.prefix() == one_prefix {
		warn!("subnet is /{}: this is almost certainly NOT what you want to do", one_prefix);
	}

	let interface = interfaces()
		.into_iter()
		.find(|i| Some(&i.name) == args.interface.as_ref())
		.ok_or_else(|| eyre!("interface does not exist"))?;
	if interface.is_loopback() {
		return Err(eyre!("cannot use loopback interface"));
	}
	if interface.is_point_to_point() {
		return Err(eyre!("cannot use point-to-point interface"));
	}
	if !interface.is_up() {
		return Err(eyre!("interface must be up"));
	}

	let mac = args
		.mac
		.or(interface.mac)
		.ok_or_else(|| eyre!("interface does not have a mac address"))?;

	Ok(Some((ip, interface, mac, !args.unmanaged_ip, args)))
}

async fn run((ip, interface, mac, mut ip_managed, args): Prep) -> Result<()> {
	let (mut tx, mut rx) = match datachannel(
		&interface,
		Config {
			channel_type: ChannelType::Layer2,
			promiscuous: true,
			..Default::default()
		},
	)? {
		Channel::Ethernet(tx, rx) => (tx, rx),
		_ => unimplemented!("internal: unhandled datachannel type"),
	};

	// TODO: use dialectic session types?
	let (oconnor, terminator) = channel::bounded(1);
	ctrlc::set_handler(move || {
		oconnor
			.try_send(())
			.expect("failed to exit, so exiting harder (unclean)");
	})?;

	let (nlconn, nl, _) = rtnetlink::new_connection()?;
	let nlah = AddressHandle::new(nl);

	if ip_managed {
		debug!("starting netlink connection");
		spawn(nlconn);

		info!("checking if interface has ip", { ip: as_display!(ip), interface: interface.index });
		if find_addr_for_ip(&nlah, interface.clone(), ip).await?.is_some() {
			if args.die_if_ip_exists {
				return Err(eyre!("ip exists on interface, abort"));
			} else {
				warn!("existing ip on the interface");
				if !args.remove_pre_existing_ip {
					ip_managed = false;
				}
			}
		} else {
			info!("adding ip to interface", { ip: as_display!(ip), interface: interface.index });
			nlah.add(interface.index, ip.ip(), ip.prefix())
				.execute()
				.await?;
		}
	}

	let (listener, blaster) = match ip {
		IpNetwork::V4(_) => {
			let (watch_signal, mut watch_pulse) = if args.watch_immediately {
				(Signal::pulsed(), None)
			} else {
				let (s, p) = Signal::new();
				(s, Some(p))
			};

			let watch = args.watch;
			let watch_delay = args.watch_delay;

			let listener = spawn_blocking(move || -> Result<()> {
				if let Watch::No = watch {
					return Ok(());
				}

				watch_signal
					.wait()
					.map_err(|_| eyre!("failed to wait on watch signal"))?;
				wait(watch_delay);

				info!("watching for competing arp announcements");

				loop {
					let pkt = rx.next()?;
					let eth = EthernetPacket::new(pkt)
						.ok_or_else(|| eyre!("eth packet buffer too small"))?;
					if eth.get_ethertype() != EtherTypes::Arp {
						continue;
					}

					let pay = eth.payload();
					let arp =
						ArpPacket::new(pay).ok_or_else(|| eyre!("arp packet buffer too small"))?;

					let op = match arp.get_operation() {
						ArpOperations::Reply => String::from("reply"),
						ArpOperations::Request => String::from("request"),
						ArpOperation(n) => format!("unknown: {}", n),
					};

					let hw = match arp.get_hardware_type() {
						ArpHardwareTypes::Ethernet => String::from("ethernet"),
						ArpHardwareType(n) => format!("unknown: {}", n),
					};

					let gratuitous = arp.get_sender_proto_addr() == arp.get_target_proto_addr();

					debug!("read arp packet", {
						src: as_display!(eth.get_source()),
						dst: as_display!(eth.get_destination()),
						op: as_display!(op),
						hw: as_display!(hw),
						proto: as_display!(arp.get_protocol_type()),
						sender_hw: as_display!(arp.get_sender_hw_addr()),
						sender_proto: as_display!(arp.get_sender_proto_addr()),
						target_hw: as_display!(arp.get_target_hw_addr()),
						target_proto: as_display!(arp.get_target_proto_addr()),
						gratuitous: gratuitous,
					});

					if gratuitous
						&& arp.get_sender_proto_addr() == ip.ip()
						&& arp.get_sender_hw_addr() != mac
					{
						match watch {
							Watch::No => unreachable!(),
							Watch::Fail => {
								return Err(eyre!(
									"received competing announce! src={} mac={}",
									eth.get_source(),
									arp.get_sender_hw_addr()
								))
							}
							Watch::Quit => {
								info!("received competing announce!", {
									src: as_display!(eth.get_source()),
									mac: as_display!(arp.get_sender_hw_addr()),
								});
								return Ok(());
							}
							Watch::Log => {
								warn!("received competing announce!", {
									src: as_display!(eth.get_source()),
									mac: as_display!(arp.get_sender_hw_addr()),
								});
							}
						}
					}
				}
			});

			let blaster = spawn_blocking(move || -> Result<()> {
				wait(args.delay);

				let mut n = 0_usize;
				loop {
					let mut arp_buf = vec![0_u8; MutableArpPacket::minimum_packet_size()];
					let mut arp = MutableArpPacket::new(&mut arp_buf[..])
						.ok_or_else(|| eyre!("failed to create arp packet"))?;

					let ip4 = match ip.ip() {
						IpAddr::V4(i) => i,
						_ => unreachable!(),
					};

					arp.set_protocol_type(EtherTypes::Ipv4);
					arp.set_hardware_type(ArpHardwareTypes::Ethernet);
					arp.set_hw_addr_len(6);
					arp.set_proto_addr_len(4);
					arp.set_sender_hw_addr(mac);
					arp.set_target_hw_addr(mac);
					arp.set_sender_proto_addr(ip4);
					arp.set_target_proto_addr(ip4);
					arp.set_operation(if args.arp_reply {
						ArpOperations::Reply
					} else {
						ArpOperations::Request
					});

					let mut eth_buf = vec![
						0_u8;
						MutableEthernetPacket::minimum_packet_size()
							+ MutableArpPacket::minimum_packet_size()
					];
					let mut eth = MutableEthernetPacket::new(&mut eth_buf)
						.ok_or_else(|| eyre!("failed to create eth packet"))?;

					eth.set_source(mac);
					eth.set_destination(args.target);
					eth.set_ethertype(EtherTypes::Arp);
					eth.set_payload(arp.packet_mut());

					info!("sending arp packet", {
						n: n,
						src: as_display!(mac),
						dst: as_display!(args.target),
						op: if args.arp_reply { "reply" } else { "request" },
						hw: "ethernet",
						hw_addr: as_display!(mac),
						proto_addr: as_display!(ip4),
						gratuitous: true,
					});
					tx.send_to(eth.packet(), None)
						.transpose()?
						.ok_or_else(|| eyre!("unknown error sending packet"))?;

					n = n.saturating_add(1);
					if args.count > 0 && n >= args.count {
						return Ok(());
					}

					if let Some(pulse) = watch_pulse.take() {
						pulse.pulse();
					}

					wait(jittered(args.interval, args.jitter));
				}
			});

			(listener, blaster)
		}
		IpNetwork::V6(_) => todo!("ipv6 support"),
	};

	if let Err(err) = terminator
		.recv()
		.map_err(|e| e.into())
		.race(listener)
		.race(blaster)
		.await
	{
		eprintln!("{:?}", err);
	}

	if ip_managed {
		info!("removing ip from interface", { ip: as_display!(ip), interface: interface.index });
		if let Some(addr) = find_addr_for_ip(&nlah, interface, ip).await? {
			nlah.del(addr).execute().await?;
		}
	}

	Ok(())
}

async fn find_addr_for_ip(
	nlah: &AddressHandle,
	interface: NetworkInterface,
	ip: IpNetwork,
) -> Result<Option<AddressMessage>> {
	let mut addrlist = nlah.get().execute();
	while let Some(addr) = addrlist.try_next().await? {
		if addr.header.index != interface.index {
			continue;
		}

		let addrbytes = match addr.nlas.iter().find(|n| matches!(n, Nla::Address(_))) {
			Some(Nla::Address(a)) => a,
			_ => continue,
		};

		match ip {
			IpNetwork::V4(ip4) => {
				match <[u8; 4]>::try_from(addrbytes.clone()).map(Ipv4Addr::from) {
					Ok(ar) if ar == ip4.ip() => return Ok(Some(addr)),
					_ => continue,
				};
			}
			IpNetwork::V6(ip6) => {
				match <[u8; 16]>::try_from(addrbytes.clone()).map(Ipv6Addr::from) {
					Ok(ar) if ar == ip6.ip() => return Ok(Some(addr)),
					_ => continue,
				};
			}
		};
	}

	Ok(None)
}
